narwhals/_pandas_like/group_by.py

```diff
@@ -40,231 +249,121 @@ def __init__(
         *,
         drop_null_keys: bool,
     ) -> None:
-        self._df = df
+        self._original_columns = tuple(df.columns)
         self._drop_null_keys = drop_null_keys
-        self._compliant_frame, self._keys, self._output_key_names = self._parse_keys(
-            df, keys=keys
+        ns = df.__narwhals_namespace__()
+        frame, self._keys, self._output_key_names = self._parse_keys(df, keys=keys)
+        self._exclude: tuple[str, ...] = (*self._keys, *self._output_key_names)
+        self._remap_non_str_columns = _remap_non_str(self)
+        self._compliant_frame = frame.with_columns(
+            *(ns.col(old).alias(new) for old, new in self._remap_non_str_columns.items())
```

Comment:
This should avoid a horizontal concat:

```suggestion
        self._remap_non_str_columns = _remap_non_str(self)
        self._compliant_frame = frame.rename(self._remap_non_str_columns)
```

Reply:
Yeah a fast path in `PandasDataFrame.rename` (and others I suppose?) sounds like a good idea to me!

There's probably a few things that'll need to be *untangled* from the current logic to make this work - but if we *can* rename instead then the end result may be simpler as well 🙂

Reply:
> Yeah a fast path in PandasDataFrame.rename (and others I suppose?) sounds like a good idea to me!

Yes the annoying part is that it would be nice and easy to do it at the Narwhals level. Yet, when we need it here, we are operating at the compliant level.

Since for now this is something specific to pandas, I would address it in:

- Narwhals level: https://github.com/narwhals-dev/narwhals/blob/6bf173100b040670fbf6823d647d968e332ea59d/narwhals/dataframe.py#L178 
  - Notice that polars does not return the original object but a copy (?) as `df is df.rename({})` is False.

- Pandas-like: https://github.com/narwhals-dev/narwhals/blob/6bf173100b040670fbf6823d647d968e332ea59d/narwhals/_pandas_like/dataframe.py#L460

---

I can take it and address it separately

Reply:
Hey @dangotbanned in scikit-lego we heavily use group_by's, so that's a good test to have. I will take a look tomorrow. Is it possible to be related with returning the same object vs a copy of it? Similar to the comment I made here: https://github.com/narwhals-dev/narwhals/pull/2680#discussion_r2193400231

Reply:
> Is it possible to be related with returning the same object vs a copy of it?

I had a similar thought in (https://github.com/narwhals-dev/narwhals/pull/2680/commits/24eb87301c07ba4255c5d0ee0090a9a84620a8cc) but I assumed these were all cases where we *actually* rename.
Whereas (https://github.com/narwhals-dev/narwhals/pull/2680#discussion_r2193400231) is for the inverse IIUC?

Reply:
Regarding returning a copy vs original object: I am not sure what's the best way of achieving it.

- `rename` allows empty mapping (and returns a copy)
- `assign` allows empty args as well (and returns a copy) (could this be the fastpath in `with_columns`? or just use `native.copy()`?)

They seems to run with similar speed:

```py
import numpy as np
import pandas as pd

frame = pd.DataFrame(np.random.randn(1_000_000, 100))

%timeit frame.rename(columns={})
458 ms ± 3.18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit frame.assign()
461 ms ± 6.79 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

frame is frame.rename(columns={})
False

frame is frame.assign()
False
```

Reply:
> Regarding returning a copy vs original object: I am not sure what's the best way of achieving it.

> They seem to run with similar speed - I assume both rename and assign without args just return a copy without other operations (aside checking the args):

I must say, I am tempted to discuss how we can optimize this 😅 

But I think the priorities we should have at the moment are:

1. Getting the [`scikit-lego` tests](https://github.com/narwhals-dev/narwhals/actions/runs/16169431080/job/45639010549?pr=2680) green again
2. Adding a *more minimal* test case to `narwhals` that would pass on `main` and in whatever fix we arrived at in step **1**
3. Removing the parts of this PR that still assume `PandasLikeGroupBy.__init__` is adding columns aliased columns
  i. This might be the issue itself, but I haven't tracked it down yet 😔 

Then we can figure out any perf bottlenecks and see if taking any of the paths you've mentioned cause anything to regress 🙂